### DISCUSSION TITLE
zen_href_link: use of $static?

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#5789] [Ideas]
https://github.com/zencart/zencart/blob/336fc4f6c9358ef8ed16db397ab0d5aaa143c230/includes/functions/html_output.php#L48-L60

In what context would this be used?

(I was looking for an inbuilt way to defeat any url-remapping and retain the native url, but this is not it)

With static as true it produces a link like this:

https://www.website/product_info?products_id=1244 

which obviously does not work.

**Comment by lat9** [comment]
It'll depend on how the URL remapping deals with the `NOTIFY_SEFU_INTERCEPT` notification.

_Ideally_, and I'm seeing that USU isn't _ideal_, an observer on that notification would look to see if the `$link` is still `null` on entry and, if so, return from the notification without modification.

That would enable you to load an observer before the main URL mangler, set the link to a non-null value and have the mangler subsequently ignore the URL-generation request.

**Comment by torvista** [comment]
I understand you are suggesting an additional observer before NOTIFY_SEFU_INTERCEPT.
I was thinking that using a parameter in the function call to indicate it should not be remapped, would be easiest to use and to see/more intuitive than a new observer which would require identifying the exact source of the call to know whether to act or not.

Hence I was looking at $static, thinking it may be already in place, and not understanding what it is currently good for?

**Comment by mc12345678** [comment]
> https://github.com/zencart/zencart/blob/336fc4f6c9358ef8ed16db397ab0d5aaa143c230/includes/functions/html_output.php#L48-L60
> 
> In what context would this be used?
> 
> (I was looking for an inbuilt way to defeat any url-remapping and retain the native url, but this is not it)
> 
> With static as true it produces a link like this:
> 
> https://www.website/product_info?products_id=1244
> 
> which obviously does not work.

Not entirely sure about what aspect is of interest in using a static link directly, but either the calling code needs to ensure that `$page` ends with `.php` or whatever extension is desired to be used or this code should force a php extension. I suggest the former. 

Otherwise, if going to use an observer to filter the output then there are (or should be) ways to set a flag on a given page where links need to be a particular way so that the observer operates as desired. Can only generalize based on the limited data provided.

**Reply by torvista** [reply]
To clarify, I understand now that a static link is a "friendly" one ("widget"), the dynamic link is the "native" one ("&product_id = 123").
When a url rewriter is in use, all the links shopfront and admin are converted to "friendly".

But, when I want to paste a link in a product description to another product, I want to copy/use the native one, so it does not get broken by some subsequent renaming of a product or category. Also, maybe there is an argument to use the native url in sitemap and search engine feeds...I really don't know/that's another discussion.

So, unless there is a mechanism to prevent url rewriting, there's no way to get at the native url in the admin.
I've added a parameter to zen_href_link so it skips the rewriter observer, and in the admin category/product listing I make the pid column to use native links to the product page, just for copying.
That's the backstory.

And so when I saw the already existing $static parameter, I wonder what is that used for?
